augur/filter.py

```diff
@@ -3,381 +3,211 @@
-    parser.add_argument('--subsample-seed', help="random number generator seed to allow reproducible sub-sampling (with same input data). Can be number or string.")
-    parser.add_argument('--exclude-where', nargs='+',
-                                help="Exclude samples matching these conditions. Ex: \"host=rat\" or \"host!=rat\". Multiple values are processed as OR (matching any of those specified will be excluded), not AND")
-    parser.add_argument('--include-where', nargs='+',
-                                help="Include samples with these values. ex: host=rat. Multiple values are processed as OR (having any of those specified will be included), not AND. This rule is applied last and ensures any sequences matching these rules will be included.")
-    parser.add_argument('--output', '-o', help="output file", required=True)
+    parser.add_argument(
+        "--sequences",
+        "-s",
+        required=True,
+        help="sequences in FASTA, VCF, or gzipped VCF format (detected by file extension)",
+    )
+    parser.add_argument(
+        "--metadata", required=True, help="metadata associated with sequences"
+    )
+    parser.add_argument("--output", "-o", required=True, help="output file")
+
+    for cls in MATCHER_CLASSES.values():
+        cls.add_arguments(parser)
 
 
 def run(args):
-    '''
+    """
     filter and subsample a set of sequences into an analysis set
-    '''
-    #Set flags if VCF
-    is_vcf = False
-    is_compressed = False
-    if any([args.sequences.lower().endswith(x) for x in ['.vcf', '.vcf.gz']]):
-        is_vcf = True
-        if args.sequences.lower().endswith('.gz'):
-            is_compressed = True
-
-    ### Check users has vcftools. If they don't, a one-blank-line file is created which
-    #   allows next step to run but error very badly.
-    if is_vcf:
-        from shutil import which
-        if which("vcftools") is None:
-            print("ERROR: 'vcftools' is not installed! This is required for VCF data. "
-                  "Please see the augur install instructions to install it.")
-            return 1
+    """
+    Filterer(args).run()
+
+
+class Filterer:
+    def __init__(self, args):
+        self.args = args
+        self.sequence_file = SequenceFile(args.sequences, args.metadata)
+
+        self.check_vcftools()
+
+        self.sequences = self.sequence_file.sequences
+
+    def run(self):
+        self.filter_sequences()
+        self.print_summary()
+        self.write_output_files()
+
+    def translate_args(self):
+        # TODO This is technical debt to accommodate the legacy command-line options.
+        # Ideally, the CLI args should all be either --exclude or --include. Something like:
+        #     --exclude names:file=names.txt --exclude date:min=2012 --include metadata:region=north_america
```

Comment:
This syntax is, of course, open to discussion.

We can even make it order-dependent, so the later directives take precedence over the earlier ones (including `--subsample`?). Give the user control over that.